### PR TITLE
[decompositions] add RNN decomp and testing

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91466
* #91465
* #91281
* #91124
* #91807
* __->__ #91123

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

### CODE REVIEW

File: test/test_decomp.py [hunk 1]
```diff
@@ -406,6 +407,111 @@ def test_uniform(self, device):
...
+            m.to(device).to(dtype)
+
+            args, kwargs = module_input.forward_input.args, module_input.forward_input.kwargs
+            with self.DecompCrossRefMode(self, self.precision, self.rel_tol, dtype), enable_python_dispatcher():
+                decomp_out = m(*args, **kwargs)
+
+            non_decomp_out = m(*args, **kwargs)
+            if is_python_dispatcher_decomp:
+                # without this check, incorrect decomps at the python dispatcher level can still pass because
+                # they're checking aten decomps at the
```
Comment (zou3519) [inline test/test_decomp.py:433]: Comment cut off -- why is is_python-dispatcher_decoomp necessary?

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -2065,6 +2065,154 @@ def upsample_nearest3d(
...
+            bwd_inp = bwd_inp.flip(0)
+            final_hiddens.append(bwd_hidden)
+
+        if bidirectional:
+            input = torch.cat([fwd_inp, bwd_inp], fwd_inp.dim() - 1)
+        else:
+            input = fwd_inp
+
+        if dropout != 0 and train and i < num_layers - 1:
+            input = torch.dropout(input, dropout, train=True)
```
Comment (zou3519) [inline torch/_decomp/decompositions.py:2150]: just to check: dropout isn't actually tested, right? I don't see it in ModuleInfo: https://github.com/pytorch/pytorch/blob/ce50a8de7535b5e359f7ed7ead4285414a966d3f/torch/testing/_internal/common_modules.py#L951

File: torch/_decomp/decompositions.py [hunk 3]
```diff
@@ -2065,6 +2065,154 @@ def upsample_nearest3d(
...
+
+    input = input.transpose(0, 1) if batch_first else input
+    return input, final_hiddens
+
+
+@aten.rnn_tanh.input.py_impl(DispatchKey.CompositeImplicitAutograd)
+@aten.rnn_tanh.input.py_impl(DispatchKey.Autograd)
+def rnn_tanh_input(
```
Comment (zou3519) [inline torch/_decomp/decompositions.py:2158]: What are the downstream ramifications of this change? e.g. does this cause PT2 to use this decomposition?

**Review by zou3519 — COMMENTED** [review]
I'm a bit confused about the testing and want to understand what the downstream ramifications of this change are (like, does it change behavior in PT2 or another subsystem, or is another change necessary to activate that?). Other than that, the RNN code looks correct to me.

File: test/test_decomp.py (hunk 1 above)
Comment (samdow) [reply test/test_decomp.py:433]: Ahhh yeah so this is what I pieced together that the testing did:
(1) runs the python dispatcher
(2) runs a mode that checks that the decompositions at the python key level match their non-decomposed versions

It doesn't check that the overall function matches the original function without the python dispatcher or any decompositions matches the decomposed version. I think this because if there's small numerical differences, they might get magnified. However, since we're mostly testing the python dispatcher decomp here (not the decomposition of linear or something), I do specifically want to check that these decompositions match. To your other points, we wouldn't need this if this was a post-autograd decomposition only

Originally I had figured that we could use the python dispatcher directly in expanded weights but since we aren't doing that I guess we could have these be post autograd decompositions instead and we wouldn't need this

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (samdow) [reply torch/_decomp/decompositions.py:2150]: Yeah this was copied from the native implementation. I can add it but I don't think the reference function supports it so it might break some tests

File: torch/_decomp/decompositions.py (hunk 3 above)
Comment (samdow) [reply torch/_decomp/decompositions.py:2158]: Yes this causes PT2 to use the decomposition. From some other issue I thought that's what we want with the theory that dynamo/inductor would re-coop performance (also since I doubt we would be able to do symbolic shapes through RNNs since we don't know the buffer size)

But since those are closed, I would be happy to make these post autograd decompositions only instead of pre autograd. I don't really like the idea of having another set of decompositions for expanded weights only in case someone needs to add this to PT2 later (it's kinda hard to discover) but I'm open hear other opinions

File: torch/_decomp/decompositions.py (hunk 2 above)
Comment (zou3519) [reply torch/_decomp/decompositions.py:2150]: No need to add testing for it, I don't think it would test correctly due to randomness especially on CUDA

File: torch/_decomp/decompositions.py (hunk 3 above)
Comment (zou3519) [reply torch/_decomp/decompositions.py:2158]: Is there a way to separate out the "adding to PT2" from this PR?

> From some other issue I thought that's what we want with the theory that dynamo/inductor would re-coop performance (also since I doubt we would be able to do symbolic shapes through RNNs since we don't know the buffer size)

Do you have a link to the issues and can we ask the folks cc'ed there for an opinion? In theory dynamo/inductor can re-coop the performance if we decompose RNNs. In practice, there's a lot they need to do to the replicate the RNN optimizations on CUDA; the cudnn kernels are highly optimized. I'd want to get some more eyes on this from people who work on PT2 perf (@ngimel?)

File: test/test_decomp.py (hunk 1 above)
Comment (zou3519) [reply test/test_decomp.py:433]: It sounds like we're worried about small numeric differences and tolerances. If that is the case, can we always test in float64 and remove the special case for `is_python_dispatcher_decomp`? The special case makes it so that the test isn't very ModuleInfo-agnostic; someone updating it in the future needs to update the condition

File: torch/_decomp/decompositions.py (hunk 3 above)
Comment (samdow) [reply torch/_decomp/decompositions.py:2158]: > Do you have a link to the issues and can we ask the folks cc'ed there for an opinion?

https://github.com/pytorch/functorch/issues/586
Animesh is out and I'm not certain if this is the right function but it looks like @ezyang might have just disallowed dynamo from capturing RNNs at all

Comment (zou3519) [reply torch/_decomp/decompositions.py:2158]: If dynamo doesn't capture RNNs then my concerns are alleviated

Comment (ezyang) [reply torch/_decomp/decompositions.py:2158]: Well, I mean, I just caused us to hit `set_`, which means they get graph broken. RNNs are complicated because of the weight sharing thing, I am pretty doubtful inductor knows how to deal with it.

**Review by zou3519 — APPROVED** [review]
Comment on testing is a minor nit, this LGTM

File: test/test_decomp.py (hunk 1 above)
Comment (samdow) [reply test/test_decomp.py:433]: Ah sorry I wasn't clear. So the issue is that small differences get magnified when they're propagated. This happens for float64 and 32 and the differences end up being rather significant (abs diff of sometimes like 1.5, relative difference of like 8). The rest of the testing has this issue but it's picked up by proxy tensor tests (they're just harder to write for modules). The errors I was seeing were all for the RNNCells. Since there aren't any decompositions for the RNNCells, it seemed more like a testing bug 

But I totally hear you on the special case, a different option that might be nicer is instead of testing all the modules only test RNNs since we've only written decomps for RNNs). It's a special case but a special case in the test filter. I just put up this change for this but can revert if the previous version is preferred
